File: R/2_DP_Discretization/main_ddp.R
# メインファイル
# 状態変数と操作変数を離散化して動的計画法(discretized DP)を解く.

# environmentから変数をクリア
rm(list=ls())

# カリブレーション
beta  <- 0.96 # 割引因子
gamma <- 1.0  # 相対的危険回避度(異時点間の代替の弾力性の逆数)
alpha <- 0.40 # 資本分配率
delta <- 1.00 # 固定資本減耗

# 離散化用のパラメータ
nk   <- 1001  # グリッドの数
kmax <- 0.5   # 資本グリッドの最大値
kmin <- 0.05  # 資本グリッドの最小値 (0にすると生産が出来なくなる)

# 収束の基準
maxit <- 1000    # 繰り返し計算の最大値
tol   <- 1.0e-005 # 許容誤差(STEP 2)
dif1  <- 1.0      # 価値関数の繰り返し誤差
dif2  <- 1.0      # 政策関数の繰り返し誤差
#----------------------------------------

# 計算開始
time_start <- proc.time()

cat('', "\n")
cat('-+- Solving an optimal growth model -+-', "\n")
cat('', "\n")

# STEP 1(a):グリッド生成
kgrid <- seq(from = kmin, to = kmax, length.out = nk)

# STEP 1(b):価値関数・政策関数の初期値を設定
vfcn  <- matrix(data = 0.0, nrow = nk, ncol = 1)
pfcn  <- matrix(data = 0.0, nrow = nk, ncol = 1)
Tvfcn <- matrix(data = 0.0, nrow = nk, ncol = 1)
Tpfcn <- matrix(data = 0.0, nrow = nk, ncol = 1)
vkp   <- matrix(data = 0.0, nrow = nk, ncol = nk)

# STEP 3: 効用関数の組み合わせ

# 効用関数の初期値 (消費が0以下になる組み合わせにはペナルティ)
util <- matrix(data = -10000.0, nrow = nk, ncol = nk)

# 消費が正値になる(k,k')の組み合わせについて効用を計算
for (i in 1:nk) {
    #  あらゆる操作変数k'について:
    for (j in 1:nk) {
        wealth <- kgrid[i]^alpha + (1.0-delta)*kgrid[i]
        cons   <- wealth - kgrid[j]
        if (cons > 0) {
            if (gamma == 1.0) {
                util[j, i] <- log(cons)
            } else {
                util[j, i] <- cons^(1.0-gamma) / (1.0-gamma)
            }
        }
    }
}

# STEP 4: 価値関数を繰り返し計算
for (it in 1:maxit) {

    #if (dif1 < tol || dif2 < tol) break
    if (dif1 < tol) break

    # ベルマン方程式: V(k;k')
    for (i in 1:nk) {
        vkp[1:nk, i] = util[1:nk, i] + beta*vfcn
    }

    # 最適化: 各kについてV(k;k')を最大にするk'を探す
    for (i in 1:nk) {
        Tvfcn[i] <- max(vkp[1:nk, i])
        Tpfcn[i] <- kgrid[which.max(vkp[1:nk, i])]
    }

    # 繰り返し計算誤差を確認
    dif1 <- max(abs((Tvfcn-vfcn)/vfcn))
    dif2 <- max(abs((Tpfcn-pfcn)/pfcn))

    # 価値関数・効用関数をアップデート
    vfcn <- Tvfcn
    pfcn <- Tpfcn
    cat(" iteration index:", it, ", iter dif of value:", dif1, ", iter dif of policy:", dif2, "\n")

}

# 計算時間をカウント終了
cat(" Time = ", proc.time() - time_start, "\n")
cat('', "\n")

# 計算結果を表示
cat('-+- Parameter values -+-')
cat(" beta = ", beta, ", gamma = ", gamma, ", alpha = ", alpha, ", delta = ", delta, "\n")
cat(" kmin:", kmin, ", kmax:", kmax, ", #grid:", nk, "\n")
cat("\n")

# 解析的解
AA = (1.0-beta)^(-1) * (log(1.0-alpha*beta) + ((alpha*beta)/(1.0-alpha*beta))*log(alpha*beta))
BB = alpha/(1.0-alpha*beta)
v_true = AA + BB*log(kgrid)
p_true = beta*alpha*(kgrid**alpha)

# 図を描く

library("ggplot2")
library("extrafont")

# Figure 1: value function
df <- data.frame(k = kgrid, v1 = vfcn, v2 = v_true)
df1 <- data.frame(cap = df$k, v1 = df$v1, group="Approx")
df2 <- data.frame(cap = df$k, v2 = df$v2, group="Analytical")

g1 <- ggplot()
g1 <- g1 + geom_line(data=df1, aes(x = cap, y = v1, color=group), size=1.5)
g1 <- g1 + geom_line(data=df2, aes(x = cap, y = v2, color=group), size=1.5, linetype="dashed")
g1 <- g1 + scale_color_manual(name="Legend", values=c("Approx"="blue", "Analytical"="red"))
g1 <- g1 + labs(title="Value Function", x="Capital: k", y="Value Function: V(k)")
plot(g1)
ggsave(file = "Fig_dndp1.eps", width = 8, height = 6)

# Figure 2: policy function
fp <- data.frame(k = kgrid, v1 = pfcn, v2 = p_true, v3 = kgrid)
fp1 <- data.frame(cap = fp$k, v1 = fp$v1, group="Approx")
fp2 <- data.frame(cap = fp$k, v2 = fp$v2, group="Analytical")
fp3 <- data.frame(cap = fp$k, v3 = fp$v3, group="45-degree")

g2 <- ggplot()
g2 <- g2 + geom_line(data=fp1, aes(x = cap, y = v1, color=group), size=1.5)
g2 <- g2 + geom_line(data=fp2, aes(x = cap, y = v2, color=group), size=1.5)
g2 <- g2 + geom_line(data=fp3, aes(x = cap, y = v3, color=group), size=1.5, linetype="dotted")
g2 <- g2 + scale_color_manual(name="Legend", values=c("Approx"="blue", "Analytical"="red", "45-degree"="black"))
g2 <- g2 + labs(title="Policy Function", x="Current Capital", y="Next Period's Capital")
plot(g2)
ggsave(file = "Fig_dndp2.eps", width = 8, height = 6)

File: Python/2_DP_Discretization/main_ddp.py
"""
メインファイル:
状態変数と操作変数を離散化して動的計画法(discretized DP)を解く.
"""

# pylint: disable=C0103

import copy
import math
import time
import numpy as np
#import matplotlib as mpl
import matplotlib.pyplot as plt

## カリブレーション
beta = 0.96 # 割引因子
gamma = 1.0 # 相対的危険回避度(異時点間の代替の弾力性の逆数)
alpha = 0.40 # 資本分配率
delta = 1.00 # 固定資本減耗(0.08)

## 離散化用のパラメータ
nk = 101 # グリッドの数
kmax = 0.5 # 資本グリッドの最大値
#kmax = 10.0 # 資本グリッドの最大値(固定資本減耗=0.08の場合に使用)
kmin = 0.05 # 資本グリッドの最小値(0にすると生産が出来なくなる)

## 収束の基準
it = 1 # ループ・カウンター
maxit = 1000 # 繰り返し計算の最大値
tol = 1.0e-005 # 許容誤差(STEP 2)
dif1 = 1.0 # 価値関数の繰り返し誤差
dif2 = 1.0 # 政策関数の繰り返し誤差

start = time.time()

print("-+- Solve a neoclassical growth model -+-")

## STEP 1(a): グリッド生成

kgrid = np.linspace(kmin, kmax, nk)

## STEP 1(b): 価値関数・政策関数の初期値を設定

vfcn = np.zeros([nk])
pfcn = np.zeros([nk])
Tvfcn = np.zeros([nk])
Tpfcn = np.zeros([nk])
vkp = np.zeros([nk, nk])

## STEP 3: 効用関数の組み合わせ

# 効用関数の初期値 (消費が0以下になる組み合わせにはペナルティ)
util = -10000.0*np.ones([nk, nk])

# 消費が正値になる(k,k')の組み合わせについて効用を計算
for i in range(nk):
    #  あらゆる操作変数k'について:
    for j in range(nk):
        wealth = kgrid[i]**alpha + (1.0-delta)*kgrid[i]
        cons = wealth - kgrid[j]
        if cons > 0:
            if gamma == 1.0:
                util[j, i] = math.log(cons)
            else:
                util[j, i] = cons**(1.0-gamma) / (1.0-gamma)

## STEP 4: 価値関数を繰り返し計算
for it in range(maxit):

    #if dif1 < tol or dif2 < tol:
    #    break
    if dif1 < tol:
        break

    # ベルマン方程式: V(k;k')
    for i in range(nk):
        vkp[0:nk, i] = util[0:nk, i] + beta*vfcn

    # 最適化: 各kについてV(k;k')を最大にするk'を探す
    for i in range(nk):
        Tvfcn[i] = max(vkp[:, i])
        Tpfcn[i] = kgrid[np.argmax(vkp[:, i])]

    # 繰り返し計算誤差を確認
    dif1 = max(abs((Tvfcn-vfcn)/Tvfcn))
    dif2 = max(abs((Tpfcn-pfcn)/Tpfcn))

    # 価値関数・政策関数をアップデート
    # in Python/Julia, "=" does not really assign the "number"
    vfcn = copy.deepcopy(Tvfcn)
    pfcn = copy.deepcopy(Tpfcn)
    print([it, dif1, dif2])

elapsed_time = time.time() - start

print('-+- Computation time -+-')
print(elapsed_time)

## 計算結果をコマンドウィンドウに表示
print('-+- Parameter Values -+-')
print('alpha  beta  gamma      ')
print([alpha, beta, gamma])
print('kmin  kmax  #grid      ')
print([kmin, kmax, nk])

## 解析的解
v_true = np.zeros([nk])
p_true = np.zeros([nk])
AA = (1.0-beta)**(-1) * (math.log(1.0-alpha*beta) + ((alpha*beta)/(1.0-alpha*beta))*math.log(alpha*beta))
BB = alpha/(1.0-alpha*beta)
for i in range(nk):
    v_true[i] = AA + BB*math.log(kgrid[i])
    p_true[i] = beta*alpha*(kgrid[i]**alpha)

## 図を描く
plt.figure()
plt.plot(kgrid, vfcn, linestyle='solid', color='blue', label='approx')
plt.plot(kgrid, v_true, linestyle='dashed', color='red', label='analytical')
plt.title("Value Function")
plt.xlabel("Capital: k")
plt.ylabel("Value Function: V(k)")
plt.legend(loc='lower right')
plt.grid(True)
plt.savefig('Fig3_dndp1.pdf')
plt.show()

plt.figure()
plt.plot(kgrid, pfcn, linestyle='solid', color='blue', label='approx')
plt.plot(kgrid, p_true, linestyle='dashed', color='red', label='analytical')
plt.plot(kgrid, kgrid, linestyle='dotted', color='black', label='45-degree')
plt.title("Policy Function")
plt.xlabel("Current Capital: k")
plt.ylabel("Next Period's Capital: V(k)")
plt.legend(loc='lower right')
plt.grid(True)
plt.savefig('Fig3_dndp2.pdf')
plt.show()
